### DISCUSSION TITLE
Unexpected error when using binary data in an update closure

### DISCUSSION THREAD

**Discussion by huntc** [nushell/nushell#13634] [Q&A] [answered]
I'm getting an unexpected error given some binary data in the pipeline. An example:

```
~> echo '{"value": "oW1PZmZzZXRVcGRhdGVkomdhZGRyZXNzGHxmb2Zmc2V0GpmsnKY="}' | from json --objects | update value {|row| $row.value | decode base64 | echo "hi"}                                                                                                                                                                                            16/08/24 16:30:30
Error:   × base64 payload isn't a valid utf-8 sequence
   ╭─[entry #261:1:76]
 1 │ echo '{"value": "oW1PZmZzZXRVcGRhdGVkomdhZGRyZXNzGHxmb2Zmc2V0GpmsnKY="}' | from json --objects | update value {|row| $row.value | decode base64 | echo "hi"} 
   ·                                                                            ────┬────
   ·                                                                                ╰── invalid utf-8 sequence of 1 bytes from index 0
   ╰────
  help: consider using the `--binary` flag
```

I believe the culprit is the `decode base64` stage. It appears to be updating back into the data type of the `value` JSON field. As that decoded data contains some non-printable characters, I'm assuming that's why there's an error. However, the pipeline isn't finished yet because of the `echo` as the last stage.

If I use a value with printable characters then all is well:

```
~> echo '{"value":"SGkgdGhlcmU="}' | from json --objects | update value {|row| $row.value | decode base64 | echo "hi"} 
╭───┬───────╮
│ # │ value │
├───┼───────┤
│ 0 │ hi    │
╰───┴───────╯
```

How might I deal with this issue? Unfortunately, the help isn't helpful either, as I'm unsure where to use the `--binary` flag. Thanks.

**Comment by ghost** [comment, marked as answer]
it usually helps to disassemble the case and get to a minimal reproduction first:  
testing the `echo` and `from json` on their own, then testing their result with the rest, then getting rid of the update, etc.  
the interesting part in this case is `"oW1PZmZzZXRVcGRhdGVkomdhZGRyZXNzGHxmb2Zmc2V0GpmsnKY=" | decode base64`

if we take a look at the `--binary` flag you already mentioned in `decode base64 --help`: `--binary - Output a binary value instead of decoding payload as UTF-8`.  
if it is not added nu will try to decode the resulting data into utf-8 (as mentioned by the error), which will not work.  
since your issue-title is `using binary data` i assume that you're aware that the data is not `utf-8` compatible.  
therefore using `---binary` is the right choice here

**Reply by huntc** [reply]
Ah I see. Apologies, I thought I was down to the minimum reproducer given the error message. Thanks for the reply and help.

**Reply by huntc** [reply]
Thanks again. All well.

**Comment by huntc** [comment]
Just to add - I'm loving nushell - thanks so much for the hard work!
